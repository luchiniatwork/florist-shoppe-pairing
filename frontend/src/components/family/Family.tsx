import React, { useEffect, useState } from "react";
import makeStyles from "@material-ui/core/styles/makeStyles";
import {gql, useQuery} from "@apollo/client";
import {
  Collapse,
  List,
  ListItem,
  ListItemIcon,
  ListItemText,
  Paper,
} from "@material-ui/core";
import { ExpandMore, ExpandLess, LocalFlorist } from "@material-ui/icons";
import IFamily from "../../types/Family";

const useStyles = makeStyles(({ spacing }) => ({
  root: {
    padding: spacing(3),
  },
}));

export const GET_FAMILIES_QUERY = gql`
  query GetFamilyAndFlowers {
    families {
      id
      name
      flowers {
        id
        name
      }
    }
  }
`;

const Family = () => {
  const classes = useStyles();
  const { loading, data } = useQuery(GET_FAMILIES_QUERY);
  const [expandedList, setExpandedList] = useState<boolean[]>([]);

  useEffect(() => {
    if (data && data.families) {
      setExpandedList(data.families.map(() => false));
    }
  }, [data]);

  const toggleExpanded = (idx: number) => {
    const newExpandedArray = [
      ...expandedList.slice(0, idx),
      !expandedList[idx],
      ...expandedList.slice(idx + 1),
    ];
    setExpandedList(newExpandedArray);
  };

  return (
    <Paper variant={"outlined"} className={classes.root}>
      <List>
        {!loading && data && Array.isArray(data.families)
          ? data.families.map((family: IFamily, idx: number) => (
              <React.Fragment key={`${idx}-family`}>
                <ListItem button onClick={() => toggleExpanded(idx)}>
                  <ListItemText primary={family.name} />
                  {expandedList[idx] ? <ExpandLess /> : <ExpandMore />}
                </ListItem>
                <Collapse in={expandedList[idx]} timeout={"auto"} unmountOnExit>
                  <List component={"div"} disablePadding>
                    {Array.isArray(family.flowers)
                      ? family.flowers.map((flower, flowerIdx) => (
                          <ListItem key={`${flowerIdx}-${idx}-family-flower`}>
                            <ListItemIcon>
                              <LocalFlorist />
                            </ListItemIcon>
                            <ListItemText primary={flower.name} />
                          </ListItem>
                        ))
                      : null}
                  </List>
                </Collapse>
              </React.Fragment>
            ))
          : "Loading..."}
      </List>
    </Paper>
  );
};

export default Family;
