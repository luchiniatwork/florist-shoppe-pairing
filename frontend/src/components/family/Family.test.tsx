import React from "react";
import {render} from 'react-dom';
import { MockedProvider } from "@apollo/client/testing";
import { act } from "react-dom/test-utils";
import Family, { GET_FAMILIES_QUERY } from "./Family";
import FamilyExmapleData from "./FamiliesExampleData.json";

const mocks = [
  {
    request: {
      query: GET_FAMILIES_QUERY,
    },
    result: { data: { families: FamilyExmapleData } },
  },
];

test("Render Family Load State", () => {
  let container = document.createElement('div');
  act(() => {
    render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Family />
      </MockedProvider>, container
    );
  });
  expect(container).toMatchSnapshot();
});

test("Render Family Final State", async () => {
  let container = document.createElement('div');
  act(() => {
    render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Family />
      </MockedProvider>, container
    );
  });
  await new Promise((resolve) => setTimeout(resolve, 0)); // wait for response
  expect(container).toMatchSnapshot();
});
