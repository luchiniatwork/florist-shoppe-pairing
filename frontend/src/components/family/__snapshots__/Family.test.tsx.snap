// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Family Final State 1`] = `
<div>
  <div
    class="MuiPaper-root makeStyles-root-1 MuiPaper-outlined MuiPaper-rounded"
  >
    <ul
      class="MuiList-root MuiList-padding"
    >
      <div
        aria-disabled="false"
        class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            Iridaceae
          </span>
        </div>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
      </div>
      <div
        aria-disabled="false"
        class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            Campanulaceae
          </span>
        </div>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
      </div>
      <div
        aria-disabled="false"
        class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            Liliaceae
          </span>
        </div>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
      </div>
    </ul>
  </div>
</div>
`;

exports[`Render Family Load State 1`] = `
<div>
  <div
    class="MuiPaper-root makeStyles-root-1 MuiPaper-outlined MuiPaper-rounded"
  >
    <ul
      class="MuiList-root MuiList-padding"
    >
      Loading...
    </ul>
  </div>
</div>
`;
